Add Check Support for Java 21 Unnamed Variables & Patterns Syntax: CatchParameterName

child of https://github.com/checkstyle/checkstyle/issues/14942
I have read check documentation: https://checkstyle.org/checks/naming/catchparametername.html#CatchParameterName
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```
PS D:\CS\test> cat src/Test.java                                                
public class Test {
    void test() {
        try {
            int x = 1/0;
        } catch (Exception _) {

        }
    }
}
PS D:\CS\test> cat config.xml                                                   
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
         <module name="CatchParameterName"/>
    </module>
</module>
PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:5:28: Name '_' must match pattern '^(e|t|ex|[a-z][a-z][a-zA-Z]+)$'. [CatchParameterName]
Audit done.
Checkstyle ends with 1 errors.
PS D:\CS\test> 
```

Describe what you expect in detail.

`CatchParameterName` should allow unnamed pattern variables by default

--- a/src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionCatchParameterNameTest.java
+++ b/src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionCatchParameterNameTest.java
@@ -40,7 +40,7 @@ public class XpathRegressionCatchParameterNameTest extends AbstractXpathTestSupp
 
     @Test
     public void testSimple() throws Exception {
-        final String pattern = "^(e|t|ex|[a-z][a-z][a-zA-Z]+)$";
+        final String pattern = "^(e|t|ex|[a-z][a-z][a-zA-Z]+|_)$";
 
         final DefaultConfiguration moduleConfig =
                 createModuleConfig(CatchParameterNameCheck.class);
@@ -65,7 +65,7 @@ public class XpathRegressionCatchParameterNameTest extends AbstractXpathTestSupp
 
     @Test
     public void testNested() throws Exception {
-        final String pattern = "^(e|t|ex|[a-z][a-z][a-zA-Z]+)$";
+        final String pattern = "^(e|t|ex|[a-z][a-z][a-zA-Z]+|_)$";
 
         final DefaultConfiguration moduleConfig =
                 createModuleConfig(CatchParameterNameCheck.class);

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/CatchParameterNameCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/CatchParameterNameCheck.java
@@ -35,6 +35,7 @@ import com.puppycrawl.tools.checkstyle.api.TokenTypes;
  * <li>allows {@code e} abbreviation (suitable for exceptions end errors)</li>
  * <li>allows {@code ex} abbreviation (suitable for exceptions)</li>
  * <li>allows {@code t} abbreviation (suitable for throwables)</li>
+ * <li>allows {@code _} for unnamed catch parameters</li>
  * <li>prohibits numbered abbreviations like {@code e1} or {@code t2}</li>
  * <li>prohibits one letter prefixes like {@code pException}</li>
  * <li>prohibits two letter abbreviations like {@code ie} or {@code ee}</li>
@@ -44,7 +45,7 @@ import com.puppycrawl.tools.checkstyle.api.TokenTypes;
  * <li>
  * Property {@code format} - Sets the pattern to match valid identifiers.
  * Type is {@code java.util.regex.Pattern}.
- * Default value is {@code "^(e|t|ex|[a-z][a-z][a-zA-Z]+)$"}.
+ * Default value is {@code "^(e|t|ex|[a-z][a-z][a-zA-Z]+|_)$"}.
  * </li>
  * </ul>
  * <p>
@@ -67,7 +68,7 @@ public class CatchParameterNameCheck extends AbstractNameCheck {
      * Creates a new {@code CatchParameterNameCheck} instance.
      */
     public CatchParameterNameCheck() {
-        super("^(e|t|ex|[a-z][a-z][a-zA-Z]+)$");
+        super("^(e|t|ex|[a-z][a-z][a-zA-Z]+|_)$");
     }
 
     @Override

--- a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/naming/CatchParameterNameCheck.xml
+++ b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/naming/CatchParameterNameCheck.xml
@@ -16,13 +16,14 @@
  &lt;li&gt;allows {@code e} abbreviation (suitable for exceptions end errors)&lt;/li&gt;
  &lt;li&gt;allows {@code ex} abbreviation (suitable for exceptions)&lt;/li&gt;
  &lt;li&gt;allows {@code t} abbreviation (suitable for throwables)&lt;/li&gt;
+ &lt;li&gt;allows {@code _} for unnamed catch parameters&lt;/li&gt;
  &lt;li&gt;prohibits numbered abbreviations like {@code e1} or {@code t2}&lt;/li&gt;
  &lt;li&gt;prohibits one letter prefixes like {@code pException}&lt;/li&gt;
  &lt;li&gt;prohibits two letter abbreviations like {@code ie} or {@code ee}&lt;/li&gt;
  &lt;li&gt;prohibits any other characters than letters&lt;/li&gt;
  &lt;/ul&gt;</description>
          <properties>
-            <property default-value="^(e|t|ex|[a-z][a-z][a-zA-Z]+)$"
+            <property default-value="^(e|t|ex|[a-z][a-z][a-zA-Z]+|_)$"
                       name="format"
                       type="java.util.regex.Pattern">
                <description>Sets the pattern to match valid identifiers.</description>

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/CatchParameterNameCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/CatchParameterNameCheckTest.java
@@ -58,7 +58,7 @@ public class CatchParameterNameCheckTest extends AbstractModuleTestSupport {
 
     @Test
     public void testDefaultConfigurationOnFileWithViolations() throws Exception {
-        final String defaultFormat = "^(e|t|ex|[a-z][a-z][a-zA-Z]+)$";
+        final String defaultFormat = "^(e|t|ex|[a-z][a-z][a-zA-Z]+|_)$";
 
         final String[] expected = {
             "25:28: " + getCheckMessage(MSG_INVALID_PATTERN, "exception1", defaultFormat),
@@ -96,4 +96,17 @@ public class CatchParameterNameCheckTest extends AbstractModuleTestSupport {
                 getPath("InputCatchParameterName3.java"), expected);
     }
 
+    @Test
+    public void testCatchParameterNameUnnamed() throws Exception {
+        final String defaultFormat = "^(e|t|ex|[a-z][a-z][a-zA-Z]+|_)$";
+
+        final String[] expected = {
+            "18:28: " + getCheckMessage(MSG_INVALID_PATTERN, "__", defaultFormat),
+            "24:28: " + getCheckMessage(MSG_INVALID_PATTERN, "_BAD", defaultFormat),
+            "27:28: " + getCheckMessage(MSG_INVALID_PATTERN, "BAD__", defaultFormat),
+        };
+
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputCatchParameterNameUnnamed.java"), expected);
+    }
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/catchparametername/InputCatchParameterNameUnnamed.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/catchparametername/InputCatchParameterNameUnnamed.java
@@ -1,0 +1,36 @@
+/*
+CatchParameterName
+format = (default)^(e|t|ex|[a-z][a-z][a-zA-Z]+|_)$
+
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.naming.catchparametername;
+
+public class InputCatchParameterNameUnnamed {
+
+    void m() {
+         try {
+        } catch (Exception _) {
+        }
+        try {
+        } catch (Exception __) { // violation, 'Name '__' must match pattern*.'
+        }
+        try {
+        } catch (Error | Exception _) {
+        }
+        try {
+        } catch (Exception _BAD) { // violation, 'Name '_BAD' must match pattern*.'
+        }
+        try {
+        } catch (Exception BAD__) { // violation, 'Name 'BAD__' must match pattern*.'
+        }
+        try {
+        } catch (Throwable _) {
+            try {
+            } catch (Throwable _) {
+            }
+        }
+    }
+}

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/catchparametername/InputCatchParameterName.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/catchparametername/InputCatchParameterName.java
@@ -1,6 +1,6 @@
 /*
 CatchParameterName
-format = (default)^(e|t|ex|[a-z][a-z][a-zA-Z]+)$
+format = (default)^(e|t|ex|[a-z][a-z][a-zA-Z]+|_)$
 
 
 */

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/catchparametername/InputCatchParameterNameSimple.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/catchparametername/InputCatchParameterNameSimple.java
@@ -1,6 +1,6 @@
 /*
 CatchParameterName
-format = (default)^(e|t|ex|[a-z][a-z][a-zA-Z]+)$
+format = (default)^(e|t|ex|[a-z][a-z][a-zA-Z]+|_)$
 
 
 */

--- a/src/xdocs-examples/java/com/puppycrawl/tools/checkstyle/checks/naming/CatchParameterNameCheckExamplesTest.java
+++ b/src/xdocs-examples/java/com/puppycrawl/tools/checkstyle/checks/naming/CatchParameterNameCheckExamplesTest.java
@@ -27,7 +27,7 @@ import com.puppycrawl.tools.checkstyle.AbstractExamplesModuleTestSupport;
 
 public class CatchParameterNameCheckExamplesTest extends AbstractExamplesModuleTestSupport {
 
-    private static final String CATCH_PARAM_NAME_PATTERN_1 = "^(e|t|ex|[a-z][a-z][a-zA-Z]+)$";
+    private static final String CATCH_PARAM_NAME_PATTERN_1 = "^(e|t|ex|[a-z][a-z][a-zA-Z]+|_)$";
     private static final String CATCH_PARAM_NAME_PATTERN_2 = "^[a-z][a-zA-Z0-9]+$";
 
     @Override

--- a/src/xdocs/checks/naming/catchparametername.xml
+++ b/src/xdocs/checks/naming/catchparametername.xml
@@ -21,6 +21,7 @@
           <li>allows <code>e</code> abbreviation (suitable for exceptions end errors)</li>
           <li>allows <code>ex</code> abbreviation (suitable for exceptions)</li>
           <li>allows <code>t</code> abbreviation (suitable for throwables)</li>
+          <li>allows <code>_</code> for unnamed catch parameters</li>
           <li>prohibits numbered abbreviations like <code>e1</code> or <code>t2</code></li>
           <li>prohibits one letter prefixes like <code>pException</code></li>
           <li>prohibits two letter abbreviations like <code>ie</code> or <code>ee</code></li>
@@ -42,7 +43,7 @@
               <td>format</td>
               <td>Sets the pattern to match valid identifiers.</td>
               <td><a href="../../property_types.html#Pattern">Pattern</a></td>
-              <td><code>&quot;^(e|t|ex|[a-z][a-z][a-zA-Z]+)$&quot;</code></td>
+              <td><code>&quot;^(e|t|ex|[a-z][a-z][a-zA-Z]+|_)$&quot;</code></td>
               <td>6.14</td>
             </tr>
           </table>

--- a/src/xdocs/checks/naming/catchparametername.xml.template
+++ b/src/xdocs/checks/naming/catchparametername.xml.template
@@ -21,6 +21,7 @@
           <li>allows <code>e</code> abbreviation (suitable for exceptions end errors)</li>
           <li>allows <code>ex</code> abbreviation (suitable for exceptions)</li>
           <li>allows <code>t</code> abbreviation (suitable for throwables)</li>
+          <li>allows <code>_</code> for unnamed catch parameters</li>
           <li>prohibits numbered abbreviations like <code>e1</code> or <code>t2</code></li>
           <li>prohibits one letter prefixes like <code>pException</code></li>
           <li>prohibits two letter abbreviations like <code>ie</code> or <code>ee</code></li>

